docs/design/copy-tracking.md

```diff
@@ -10,17 +10,613 @@ results of rebasing across remote copies.
+ID would still have been clear.
+
+#### Example: Rebasing
+
+```
+$ jj log
+C rename bar->baz
+|
+B rename foo->bar
+|
+A add foo
+$ jj rebase -r C -d A
+```
+
+
+```
+$ jj log
+C rename foo->baz
+|
+| B rename foo->bar
+|/
+A add foo
+$ jj rebase -r C -d B
+```
+
+
+#### Example: Divergent renames
+
+Consider this "divergent rename" scenario:
+```
+$ jj log
+C rename foo->baz
+|
+| B rename foo->bar
+|/
+A add foo
+$ jj new B C
+```
+
+In this scenario, the regular 3-way merge of the trees without considering copy
+info results in a tree without conflicts. However, the user might reasonably
+expect to have to choose between the `bar` and `baz` names. Here's what Git says
+in this scenario:
+
+```
+$ git merge main
+CONFLICT (rename/rename): foo renamed to baz in HEAD and to bar in main.
+Automatic merge failed; fix conflicts and then commit the result.
+
+$ git st
+HEAD detached from ab0b8e3
+You have unmerged paths.
+  (fix conflicts and run "git commit")
+  (use "git merge --abort" to abort the merge)
+
+Unmerged paths:
+  (use "git add/rm <file>..." as appropriate to mark resolution)
+        added by them:   bar
+        added by us:     baz
+        both deleted:    foo
+```
+
+Interestingly, Git seems to represent this state by using index states that
+would not normally end up in the index as a result of conflicts.
+
+Here's what Mercurial says:
+
+```
+$ hg merge main
+note: possible conflict - foo was renamed multiple times to:
+ bar
+ baz
+1 files updated, 0 files merged, 0 files removed, 0 files unresolved
+(branch merge, don't forget to commit)
+```
+
+Mercurial doesn't have a place to record this state, so it just prints that
+note and leaves it at that.
+
+The model and algorithm described in this document would result in a conflict
+in the copy ID at both paths after propagating the renames.
+
+#### Example: @jonathantanmy's test case:
+
+TODO: fill this out
+
+```
+$ jj log
+E baz="baz" (resolves conflict)
+|
+D <conflict>
+|\
+C | rename bar->baz
+| |
+| B rename foo->baz
+|/
+A add foo="foo" and bar="bar"
+$ jj rebase -r E -d C
+$ jj rebase new D E -m F
+```
+
+If F is empty (auto-merged), it should have the same state as E before.
+
+
+### Log
+
+The copy graph contains all past paths and copy IDs of a file, so when doing
+`jj log <filename>`, we might want to translate that to a revset that's similar
+to `files()` but matches specific (path, copy ID) pairs instead of specific
+paths.
+
+### Annotate
+
+
+### Representation in Git
+
+Do we ever want to record renames in the Git backend? If we do, we would
+presumably store it outside the Git object, similar to how we store the change
+id for commits.
+
+What do we use for trees where we *don't* have any copy graph recorded? If we
+simply create a new copy graph based on the current path, then the caller will
+never find any copies. Do we need an indexing pass to detect all renames in a
+repo when running `jj git init`? That can be very expensive for large repos.
```

Comment:
I think doing it on `jj git init` or on repository upgrade will be a large regression to the users in the big repositories I mention in my next comment. So there could be a way to "chunk" the import with some kind of background job system which incrementally indexes the renames (which makes the bar to entry even higher, as I don't think it will be a small effort).

Reply:
Yes, background-indexing is one possibility. I added a little paragraph about that.